syncer/syncer.go

```diff
@@ -0,0 +1,285 @@
+package syncer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/0xPolygon/polygon-edge/helper/progress"
+	"github.com/0xPolygon/polygon-edge/network/event"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/hashicorp/go-hclog"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const (
+	syncerLoggerName = "syncer"
+	SyncerProto      = "/syncer/0.2"
+)
+
+var (
+	errTimeout = errors.New("timeout awaiting block from peer")
+)
+
+// XXX: Don't use this syncer for the consensus that may cause fork.
+// This syncer doesn't assume fork. Consensus may be broken.
+// TODO: Add extensibility for fork before merge
+type syncer struct {
+	logger          hclog.Logger
+	blockchain      Blockchain
+	syncProgression Progression
+
+	peerMap         *PeerMap
+	syncPeerService SyncPeerService
+	syncPeerClient  SyncPeerClient
+
+	// Timeout for syncing a block
+	blockTimeout time.Duration
+
+	// Channel to notify WatchSync that a new status arrived
+	newStatusCh chan struct{}
+}
+
+func NewSyncer(
+	logger hclog.Logger,
+	network Network,
+	blockchain Blockchain,
+	blockTimeout time.Duration,
+) Syncer {
+	return &syncer{
+		logger:          logger.Named(syncerLoggerName),
+		blockchain:      blockchain,
+		syncProgression: progress.NewProgressionWrapper(progress.ChainSyncBulk),
+		syncPeerService: NewSyncPeerService(network, blockchain),
+		syncPeerClient:  NewSyncPeerClient(logger, network, blockchain),
+		blockTimeout:    blockTimeout,
+		newStatusCh:     make(chan struct{}),
+		peerMap:         new(PeerMap),
+	}
+}
+
+// Start starts goroutine processes
+func (s *syncer) Start() error {
+	if err := s.syncPeerClient.Start(); err != nil {
+		return err
+	}
+
+	s.syncPeerService.Start()
+
+	go s.initializePeerMap()
+	go s.startPeerConnectionEventProcess()
+
+	return nil
+}
+
+// Close terminates goroutine processes
+func (s *syncer) Close() {
+	close(s.newStatusCh)
+}
+
+// initializePeerMap fetches peer statuses and initializes map
+func (s *syncer) initializePeerMap() {
+	peerStatuses := s.syncPeerClient.GetConnectedPeerStatuses()
+	s.peerMap.PutPeers(peerStatuses)
+
+	for peerStatus := range s.syncPeerClient.GetPeerStatusUpdateCh() {
+		s.peerMap.Put(peerStatus)
+
+		select {
+		case s.newStatusCh <- struct{}{}:
+		default:
+		}
+	}
+}
+
+// startPeerConnectionEventProcess processes peer connection change events
+func (s *syncer) startPeerConnectionEventProcess() {
```

Comment:
Naming suggestion: `handlePeerConnections`

Reply:
I think this name sounds a handler. This method starts new process